doc/extensibility/dataflow_best_practices.md

```diff
@@ -0,0 +1,112 @@
+
+#### Use the Fault Handler Service - `IProjectFaultHandlerService`
+The fault handler service monitors exceptions and displays the yellow notification bar when exceptions are encountered. However, it only does that when it is registered.
+
+When a dataflow block encounters an exception, it switches to a `Faulted` state and stops processing any data.
+
+If the fault handler is not registered, these exceptions will be unnoticed and, because dataflow blocks stop processing data, can lead to features that silently stop working or deadlocks that are difficult to investigate.
+
+The fault handler should be registered for the last block in the chain (`C` in this case). That is because the fault state propagates through the data flow chain, so monitoring the last block is sufficient to handle exceptions in the entire chain.
+```CSharp  
+this.FaultHandlerService.RegisterFaultHandler(
+    this.blockC.Completion,
+	CommonProjectSystemTools.DefaultReportSettings,
+	project: this.UnconfiguredProject);
+```
+#### Custom Block Completion Logic
+If needed, custom block completion logic should be chained to the last block. That is because the completion propagates through the dataflow chain.
+
+```CSharp
+this.blockC.Completion.ContinueWith(
+    _ =>
+    {
+        // Custom completion handling (e.g. dispose some internal objects)
+    },
+    CancellationToken.None,
+    TaskContinuationOptions.OnlyOnFaulted,
+    TaskScheduler.Default);
+```
+
+#### Specify `PropagateCompletion = true` when creating the link
+
+```CSharp
+this.firstLink = this.Source.LinkTo(this.blockA, new DataflowLinkOptions() { PropagateCompletion = true });
+```
+
+#### You should hold a reference to the first link (Source -> A) for later disposal
+
+Typically, there is no need to keep references to the other links, which helps reduce memory consumption.
+
+```CSharp
+this.firstLink = this.Source.LinkTo(this.blockA, new DataflowLinkOptions() { PropagateCompletion = true });
+```
+
+#### You should hold a reference for the first block (`A`) for later completion (if necessary)
+Typically, block completion is not needed and requires special attention.
+
+Usually, there is no need to keep references to the other blocks, which helps reduce memory consumption.
+
+#### Use nameFormat to specify a readable name for your block
+The nameFormat gets associated with the data flow block, and it makes it easier to identify blocks when debugging issues.
+
+Some special consideration may be needed if there is a large number of blocks being created, as it may increase memory usage.
+```CSharp
+this.blockA = DataflowBlockSlim.CreateBroadcastBlock<...>(
+    nameFormat: "Block A {1}",
+    skipIntermediateInputData: true);
+```
+
+#### Consider specifying `skipIntermediateInputData`/`skipIntermediateOutputData = true` 
+
+This improves performance, but can be used only if the latest data is needed.
```

Comment:
Maybe mention something about the pitfalls of skipping values in inputs to `SyncLink` blocks.

Reply:
That's a good idea. Do you have any suggestions about the pitfalls of using them? I am not aware of a specific issue with SyncLink, but I could think of potential problems with that.
Looping in @lifengl